### PR TITLE
Ensure that we process symbol declared events for partial symbols for…

### LINKED ISSUE #3748: Report Diagnostics in Partial Classes
I am seeing an issue when reporting diagnostics on partial classes.  I posted on [stackoverflow](http://stackoverflow.com/q/30856482/573218), but got no answers. I did get verification that others are seeing it and it was suggested that I post an issue here.

I am modifying the default analyzer project that comes from the code analyzer template to try and get it to report at all of the declarations for a partial class.

I have modified the code to:

    public override void Initialize(AnalysisContext context)
    {
        context.RegisterSymbolAction(AnalyzeSymbol, SymbolKind.NamedType);
    }

    private static void AnalyzeSymbol(SymbolAnalysisContext context)
	{
	    var namedTypeSymbol = (INamedTypeSymbol)context.Symbol;
	
	    // Find just those named type symbols with names containing lowercase letters.
	    if (namedTypeSymbol.Name.ToCharArray().Any(char.IsLower))
	    {
	        foreach (var location in namedTypeSymbol.Locations)
	        {
	            // For all such symbols, produce a diagnostic.
	            var diagnostic = Diagnostic.Create(Rule, location, namedTypeSymbol.Name);
	            context.ReportDiagnostic(diagnostic);
	        }
	    }
	}
In two separate files, I have partial classes like this:

    // File1.cs
	partial class Foo
	{
	    public string BarString;
	}

	// File2.cs
	partial class Foo
	{
	    public string FooBarString;
	}

I put breakpoints on the ReportDiagnostic and am seeing it called for each location, but within Visual Studio it only reports diagnostics within a single file.

If I put multiple implementations of Foo in a single file (and it happens to be reporting on that files declaration) then I will see both diagnostics reported.

Am I misunderstanding how diagnostics are supposed to be reported or is this a bug? If it is a bug, is it a Roslyn problem or is it a problem with Visual Studio's consumption of Roslyn?

### PR DESCRIPTION
**Pull request by mavasani**
… all analyzers.

Fixes #3748

A regression test has been added for this bug.

### CODE REVIEW

**Comment by mavasani** [comment]
@srivatsn @heejaechang can you please review?

**Comment by mavasani** [comment]
@dotnet-bot retest prtest/win/tst/unit64 please

**Comment by mavasani** [comment]
ping..

**Comment by mavasani** [comment]
@dotnet/roslyn-infrastructure can the prtest/win/tst/unit64 failure be ignored?

**Comment by mavasani** [comment]
FYI @dotnet/roslyn-compiler as the code change is in the CodeAnalysis layer. However, note that this change only affects analyzer execution in live analysis, batch compilation scenario is not affected.

**Comment by cston** [comment]
LGTM

**Comment by srivatsn** [comment]
:+1:

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisState.cs [hunk 1]
```diff
@@ -234,6 +234,7 @@ private void OnCompilationEventsGenerated_NoLock(ImmutableArray<CompilationEvent
             AddToEventsMap_NoLock(compilationEvents, filterTreeOpt);
 
             // Mark the events for analysis for each analyzer.
+            List<ISymbol> newPartialSymbols = null;
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisState.cs]: Usually we use pooled ArrayBuilders. Is there a specific reason to use List<T> here?

Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisState.cs]: Not really, I can change it.

**Comment by AlekseyTs** [comment]
LGTM, modulo the List usage.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisState.cs (hunk 1 above)
Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisState.cs]: Fixed in the new iteration.

**Comment by AlekseyTs** [comment]
LGTM

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisState.cs [hunk 2]
```diff
@@ -248,11 +249,18 @@ private void OnCompilationEventsGenerated_NoLock(ImmutableArray<CompilationEvent
                         _pooledEventsWithAnyActionsSet.Add(compilationEvent);
 
                         var symbolDeclaredEvent = compilationEvent as SymbolDeclaredCompilationEvent;
-                        if (symbolDeclaredEvent?.DeclaringSyntaxReferences.Length > 1 &&
-                            !_partialSymbolsWithGeneratedSourceEvents.Add(symbolDeclaredEvent.Symbol))
+                        if (symbolDeclaredEvent?.DeclaringSyntaxReferences.Length > 1)
                         {
-                            // already processed.
-                            continue;
+                            if (_partialSymbolsWithGeneratedSourceEvents.Contains(symbolDeclaredEvent.Symbol))
```
Comment (gafter) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisState.cs]: Concurrency bug here. What happens if two threads do this at the same time (on different parts of the partial). Won't you still get duplicate events?

Isn't this just compensating for a bug in the compiler? Shouldn't we just fix that instead?

**Pull request merged**